### DISCUSSION TITLE
Tensor Seed Code - Chapter 00 Exercise #5

### DISCUSSION THREAD

**Discussion by Shlandau** [mrdbourke/pytorch-deep-learning#1339] [Q&A] [answered]
# This is a two part question:
## **Code**
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

random_seed = 1234

torch.cuda.manual_seed(random_seed)
random_tensor1 = torch.rand([7,7])
print(f"Random Tensor 1 shape: {random_tensor1.shape}\n{random_tensor1}\n")

torch.cuda.manual_seed(random_seed)
random_tensor2 = torch.rand([1,7])
print(f"Randome Tensor 2 shape: {random_tensor2.shape}\n{random_tensor2}\n")

torch.cuda.manual_seed(random_seed)
random_tensor3 = torch.matmul(random_tensor1, random_tensor2.T)
print(f"Randome Tensor 3 shape: {random_tensor3.shape}\n{random_tensor3}\n")

## **Output**
cuda
Random Tensor 1 shape: torch.Size([7, 7])
tensor([[0.2704, 0.8700, 0.7106, 0.4166, 0.5257, 0.1977, 0.8960],
        [0.9720, 0.5379, 0.2234, 0.2099, 0.3191, 0.8466, 0.1083],
        [0.0941, 0.6255, 0.5202, 0.8134, 0.9122, 0.1853, 0.2636],
        [0.1505, 0.7793, 0.3426, 0.7177, 0.9315, 0.6852, 0.3973],
        [0.0610, 0.7642, 0.8999, 0.5447, 0.9240, 0.1630, 0.2498],
        [0.6407, 0.4373, 0.2005, 0.0369, 0.2355, 0.8386, 0.5932],
        [0.0657, 0.0010, 0.4417, 0.9607, 0.5548, 0.6991, 0.4162]])

Randome Tensor 2 shape: torch.Size([1, 7])
tensor([[0.7342, 0.1234, 0.6096, 0.5621, 0.1069, 0.1711, 0.5764]])

Randome Tensor 3 shape: torch.Size([7, 1])
tensor([[1.5796],
        [1.2756],
        [1.2017],
        [1.2646],
        [1.2644],
        [1.1779],
        [1.2764]])

 ## Part 2 - Why is the below code, even though attempting to use the same seed, resulting in different tensor values?

## **Code**
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

if device == "cpu":
  random_seed = 0
  random_seed = torch.manual_seed(random_seed)
else:
  random_seed = 1234
  random_seed = torch.cuda.manual_seed(random_seed)

random_seed
random_tensor1 = torch.rand([7,7])
print(f"Random Tensor 1 shape: {random_tensor1.shape}\n{random_tensor1}\n")

random_seed
random_tensor2 = torch.rand([1,7])
print(f"Randome Tensor 2 shape: {random_tensor2.shape}\n{random_tensor2}\n")

random_seed
random_tensor3 = torch.matmul(random_tensor1, random_tensor2.T)
print(f"Randome Tensor 3 shape: {random_tensor3.shape}\n{random_tensor3}\n")

## **Output**
cuda
Random Tensor 1 shape: torch.Size([7, 7])
tensor([[0.5088, 0.7166, 0.1067, 0.8117, 0.6750, 0.9624, 0.0378],
        [0.3281, 0.6196, 0.5666, 0.2355, 0.7474, 0.1022, 0.4730],
        [0.0703, 0.0916, 0.7375, 0.7733, 0.2358, 0.0112, 0.9984],
        [0.9628, 0.9084, 0.8448, 0.8682, 0.9891, 0.2943, 0.4227],
        [0.8447, 0.7736, 0.1943, 0.9291, 0.2563, 0.2468, 0.4842],
        [0.7576, 0.6634, 0.4110, 0.4581, 0.7600, 0.8938, 0.0404],
        [0.6330, 0.7260, 0.5212, 0.0137, 0.1391, 0.5019, 0.4210]])

Randome Tensor 2 shape: torch.Size([1, 7])
tensor([[0.3573, 0.7314, 0.6552, 0.3772, 0.5826, 0.3697, 0.1584]])

Randome Tensor 3 shape: torch.Size([7, 1])
tensor([[1.8371],
        [1.5786],
        [1.1667],
        [2.6414],
        [1.6626],
        [1.9776],
        [1.4371]])

**Comment by leonardosalasd** [comment]
Hey! I can clear this up for you. Both issues come down to how PyTorch handles devices and basic Python syntax.

**Part 1: The Device Mismatch**
You are seeding the GPU using `torch.cuda.manual_seed()`, but `torch.rand()` creates tensors on the **CPU** by default. Because CPU and GPU have completely separate random number generators, your CUDA seed isn't affecting those tensors at all. 
To fix this, you have to explicitly tell the tensor to use the GPU: `torch.rand([7,7], device=device)`. 
*(Pro-tip: Just use `torch.manual_seed(1234)` instead. In modern PyTorch, that function automatically seeds both the CPU and all available GPUs for you).*

**Part 2: The Missing Function Call**
In Part 2, just typing the variable name `random_seed` on an empty line doesn't actually do anything in Python. It doesn't re-seed the generator. Because the seed was only set once at the very top, the RNG isn't restarting; it's just giving you the *next* random numbers in the sequence.
If you want to reset the sequence to get the exact same numbers again, you have to literally call the function `torch.manual_seed(random_seed)` again before every `torch.rand()` call.

**Reply by Shlandau** [reply]
Awesome! Thank you so much for the prompt and thorough response!

**Reply by Shlandau** [reply]
I re-wrote the code as follows:

## Code
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

if device == "cpu":
  random_seed = 0
else:
  random_seed = 1234

torch.manual_seed(random_seed)
random_tensor1 = torch.rand([7,7], device=device)
print(f"Random Tensor 1 shape: {random_tensor1.shape}\n{random_tensor1}\n")

torch.manual_seed(random_seed)
random_tensor2 = torch.rand([1,7], device=device)
print(f"Randome Tensor 2 shape: {random_tensor2.shape}\n{random_tensor2}\n")

torch.manual_seed(random_seed)
random_tensor3 = torch.matmul(random_tensor1, random_tensor2.T)
print(f"Randome Tensor 3 shape: {random_tensor3.shape}\n{random_tensor3}\n")

## Output - CPU
cpu
Random Tensor 1 shape: torch.Size([7, 7])
tensor([[0.0290, 0.4019, 0.2598, 0.3666, 0.0583, 0.7006, 0.0518],
        [0.4681, 0.6738, 0.3315, 0.7837, 0.5631, 0.7749, 0.8208],
        [0.2793, 0.6817, 0.2837, 0.6567, 0.2388, 0.7313, 0.6012],
        [0.3043, 0.2548, 0.6294, 0.9665, 0.7399, 0.4517, 0.4757],
        [0.7842, 0.1525, 0.6662, 0.3343, 0.7893, 0.3216, 0.5247],
        [0.6688, 0.8436, 0.4265, 0.9561, 0.0770, 0.4108, 0.0014],
        [0.5414, 0.6419, 0.2976, 0.7077, 0.4189, 0.0655, 0.8839]])

Randome Tensor 2 shape: torch.Size([1, 7])
tensor([[0.0290, 0.4019, 0.2598, 0.3666, 0.0583, 0.7006, 0.0518]])

Randome Tensor 3 shape: torch.Size([7, 1])
tensor([[0.8613],
        [1.2761],
        [1.1540],
        [1.0134],
        [0.6783],
        [1.1122],
        [0.7266]])

## Output - GPU
cuda
Random Tensor 1 shape: torch.Size([7, 7])
tensor([[0.1272, 0.8167, 0.5440, 0.6601, 0.2721, 0.9737, 0.3903],
        [0.3394, 0.5451, 0.7312, 0.3864, 0.5959, 0.7578, 0.2126],
        [0.7198, 0.9845, 0.5518, 0.0981, 0.0582, 0.5839, 0.1083],
        [0.9461, 0.3170, 0.8328, 0.6676, 0.2886, 0.9022, 0.8115],
        [0.1784, 0.9534, 0.1486, 0.3882, 0.7977, 0.1752, 0.5777],
        [0.1949, 0.8499, 0.3125, 0.2156, 0.0383, 0.4934, 0.3138],
        [0.3121, 0.5664, 0.1266, 0.7097, 0.0040, 0.5147, 0.2811]],
       device='cuda:0')

Randome Tensor 2 shape: torch.Size([1, 7])
tensor([[0.1272, 0.8167, 0.5440, 0.6601, 0.2721, 0.9737, 0.3903]],
       device='cuda:0')

Randome Tensor 3 shape: torch.Size([7, 1])
tensor([[2.5893],
        [2.1242],
        [1.8872],
        [2.5468],
        [1.7516],
        [1.6445],
        [1.6517]], device='cuda:0')

**Reply by leonardosalasd** [reply, marked as answer]
No problem at all, and sorry for the delay in getting back to you!

Your updated code looks spot on. You can now clearly see that the re-seeding is working as intended: Random Tensor 2 now perfectly matches the first row of Random Tensor 1 on both devices. This confirms that the RNG (Random Number Generator) is restarting the sequence exactly where you want it.

The fact that CPU and GPU outputs differ between each other is also expected behavior, as they implement different underlying RNG algorithms, but your results are now 100% reproducible within each environment.

If this cleared everything up for you, feel free to mark this as the answer so it can help other people facing similar device or seeding issues in the future.

If you have any more questions or run into another roadblock, just let me know. Happy coding!

**Reply by Shlandau** [reply]
Thank you so much!
